### PR TITLE
[core] Add call_site to Tasks and Actors.

### PR DESCRIPTION
**Pull request by rynewang**
Adds a new field `string call_site` to TaskSpec. It's populated from language frontend (`_raylet.pyx`), if `RAY_record_task_actor_creation_sites` is set to `true`. The field is propagated to `TaskEvent` and `ActorTableData` and through Dashboard APIs.

Users can use `ray list task --detail` as well as the Web UI. It works for Tasks, Actors (creation) and Actor Methods.

The flag `RAY_record_task_actor_creation_sites` is disabled by default, user can enable it if they want; so by default there should be no performance costs.

Screenshots:

- Task
image

- Actor
image

- Actor Method
image

### CODE REVIEW

**Comment by Moonquakes** [comment]
Hi @rynewang , it would be great to print out the parameters received by this ray task/actor!

**Review by alanwguo — COMMENTED** [review]
Dashboard UI code lgtm.

**Review by jjyao — COMMENTED** [review]
Nice. Could you add public doc about this feature?

**Comment by rynewang** [comment]
Q: do we want to default enable this feature?

**Comment by jjyao** [comment]
> Q: do we want to default enable this feature?

No. It will use more memory.

**Comment by Moonquakes** [comment]
> No. It will use more memory.

May I know how much more memory will be consumed after turning on this feature? @jjyao

**Review by MengjinYan — COMMENTED** [review]
Only minor comment on the code logic. 

At the same time, wondering if it is possible to add a test for setting the stacktrace in addition to the manual tests?

File: src/ray/protobuf/common.proto
Comment (rynewang) [reply src/ray/protobuf/common.proto:599]: yes but I am not familiar with that API. let's do this in another PR.

**Comment by rynewang** [comment]
> > No. It will use more memory.
> 
> May I know how much more memory will be consumed after turning on this feature? @jjyao

We did not profile it, you can have a try after it's merged. Target end of this week.

File: src/ray/protobuf/common.proto
Comment (MengjinYan) [reply src/ray/protobuf/common.proto:599]: Can we create an issue for that so that the item can be tracked?

**Review by MengjinYan — COMMENTED** [review]
The core logic & tests looks good to me!

File: doc/source/ray-core/actors.rst
Comment (angelinalg) [inline doc/source/ray-core/actors.rst:485]:
```suggestion
Stacktrace invocation 
```

File: doc/source/ray-core/actors.rst
Comment (angelinalg) [inline doc/source/ray-core/actors.rst:488]:
```suggestion
Ray can optionally capture and display the stacktrace of where your code creates actors and invokes actor tasks. This feature can help with debugging and understanding the execution flow of your application.
```

File: doc/source/ray-core/actors.rst
Comment (angelinalg) [inline doc/source/ray-core/actors.rst:492]:
```suggestion
- Ray captures the stacktrace when creating actors or calling actor methods
```

File: doc/source/ray-core/actors.rst
Comment (angelinalg) [inline doc/source/ray-core/actors.rst:493]:
```suggestion
- The stacktrace is visible in:
```

File: doc/source/ray-core/actors.rst
Comment (angelinalg) [inline doc/source/ray-core/actors.rst:518]:
```suggestion
    # Capture stacktrace upon actor creation.
```

File: doc/source/ray-core/actors.rst
Comment (angelinalg) [inline doc/source/ray-core/actors.rst:521]:
```suggestion
    # Capture stacktrace upon method invocation.
```

File: doc/source/ray-core/actors.rst
Comment (angelinalg) [inline doc/source/ray-core/actors.rst:524]:
```suggestion
The stacktrace shows the exact line numbers and call stack where the actor was created and methods were invoked.
```

File: doc/source/ray-core/tasks.rst
Comment (angelinalg) [inline doc/source/ray-core/tasks.rst:302]:
```suggestion
Stacktrace invocation
```

File: doc/source/ray-core/tasks.rst
Comment (angelinalg) [inline doc/source/ray-core/tasks.rst:305]:
```suggestion
Ray can optionally capture and display the stacktrace of where your code invokes tasks. This feature can help with debugging and understanding the execution flow of your application.
```

File: doc/source/ray-core/tasks.rst
Comment (angelinalg) [inline doc/source/ray-core/tasks.rst:309]:
```suggestion
- Ray captures the stacktrace when creating and invoking tasks
```

File: doc/source/ray-core/tasks.rst
Comment (angelinalg) [inline doc/source/ray-core/tasks.rst:310]:
```suggestion
- The stacktrace is visible in:
```

File: doc/source/ray-core/tasks.rst
Comment (angelinalg) [inline doc/source/ray-core/tasks.rst:330]:
```suggestion
    # Capture the stacktrace upon task invocation
```

File: doc/source/ray-core/tasks.rst
Comment (angelinalg) [inline doc/source/ray-core/tasks.rst:334]:
```suggestion
The stacktrace shows the exact line numbers and call stack where the task was created and invoked.
```

**Review by angelinalg — COMMENTED** [review]
Just some style nits.

**Review by angelinalg — APPROVED** [review]
Approving so I don't block you, but I would appreciate if you looked at my comments. Thanks!

File: doc/source/ray-core/tasks.rst
Comment (jjyao) [inline doc/source/ray-core/tasks.rst:336]: Instead of repeating the basic same thing for actor and task, I think the better place to have this is under https://docs.ray.io/en/latest/ray-observability/user-guides/debug-apps/index.html

File: src/ray/protobuf/gcs.proto
Comment (jjyao) [inline src/ray/protobuf/gcs.proto:159]: which flag

Comment (rynewang) [reply src/ray/protobuf/gcs.proto:159]: added

File: doc/source/ray-core/tasks.rst
Comment (rynewang) [reply doc/source/ray-core/tasks.rst:336]: good point, moved here

File: cpp/src/ray/runtime/task/native_task_submitter.cc
Comment (rynewang) [reply cpp/src/ray/runtime/task/native_task_submitter.cc]: done

**Review by jjyao — COMMENTED** [review]
LG.

Can you do a global replacement of invocation stacktrace to call site, just to be consistent with object.

**Comment by rynewang** [comment]
renamed to call_site.

File: doc/source/ray-observability/user-guides/debug-apps/general-debugging.rst
Comment (scottsun94) [inline doc/source/ray-observability/user-guides/debug-apps/general-debugging.rst:238]: How about task details?

Comment (rynewang) [reply doc/source/ray-observability/user-guides/debug-apps/general-debugging.rst:238]: done

**Review by jjyao — APPROVED** [review]

**Pull request merged**
